src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -108,56 +137,84 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
             properties = selectedInstanceProps.toIndexedSeq.filter(p =>
               !p.startsWith("node.") && !p.startsWith("edge.") && !p.startsWith("metadata.") &&
                 p != "startNode" && p != "endNode" && p != "space" && p != "externalId" && p != "type")
-        ))
+        )
+      )
       .toSeq
 
-    val instanceSyncRequest = InstanceSyncRequest(
+    InstanceSyncRequest(
       `with` = Map("sync" -> tableExpression),
-      cursors = if (cursor.nonEmpty) Some(Map("sync" -> cursor)) else None,
+      cursors = currentCursor.flatMap(c => Some(Map("sync" -> c))),
       select = Map("sync" -> SelectExpression(sources = sourceReference)),
       includeTyping = Some(true)
     )
+  }
 
-    Seq(syncOut(instanceSyncRequest, selectedColumns))
+  private def syncOut(
+      hasDataFilter: Boolean,
+      filters: Array[Filter],
+      selectedColumns: Array[String],
+      currentCursor: Option[String])(
+      implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] = {
+
+    val synced: IO[(ItemsWithCursor[InstanceDefinition], Boolean)] =
+      syncWithCursor(hasDataFilter, filters, selectedColumns, currentCursor)
+    val stream: Stream[IO, ProjectedFlexibleDataModelInstance] = Stream.eval(synced).flatMap {
+      case (ItemsWithCursor(items, nextCursor), continueWithHasData) =>
+        val projectedItems =
+          items.filter(clientsideHasDataFilter).map(toProjectedInstance(_, nextCursor, selectedColumns))
+
+        val nextStream = nextCursor match {
+          case Some(c) if items.nonEmpty && items.last.lastUpdatedTime < terminationTimeStamp =>
```

Comment:
could `items` still be empty due to client side filter but that won't be the end of sync stream? Maybe nonEmpty needs to be applied to unfiltered server response